### PR TITLE
Comprehensive health check improvements

### PR DESCRIPTION
**Pull request by achave11**

### CODE REVIEW

File: src/azul/__init__.py [hunk 1]
```diff
@@ -273,6 +282,19 @@ def token_queue_name(self):
...
+    def fail_queue_name(self):
+        return config.qualified_resource_name('fail')
+
+    @property
+    def fail_fifo_queue_name(self):
+        return config.qualified_resource_name('fail', suffix='.fifo')
+
+    @property
+    def all_queue_names(self):
+        return self.token_queue_name, self.document_queue_name, self.fail_queue_name, \
```
Comment (hannes-ucsc) [inline src/azul/__init__.py]: Ding, ding, ding, ding, ding!

File: test/health_check_test_case.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+        self.assertEqual({
+            'up': True,
+            'elastic_search': {
+                'up': True
+            },
+            'queues': {
+                'up': True,
+                'azul-documents-dev': {
+                    'up': True,
+                    'messages': {
```
Comment (hannes-ucsc) [inline test/health_check_test_case.py]:
```py
'messages' : {'delayed': 0, 'invisible': 0, 'queued': 0}
```

Here and for all other leaf dictionaries.

File: src/azul/__init__.py [hunk 3]
```diff
@@ -129,11 +129,20 @@ def unqualified_resource_name_or_none(self, qualified_resource_name: str, suffix
...
-        return config.subdomain(lambda_name) + "." + config.domain_name
+    def api_lambda_domain(self, lambda_name: str) -> str:
+        return self.subdomain(lambda_name) + "." + self.domain_name
 
-    def service_endpoint(self):
-        return "https://" + config.api_lambda_domain('service')
+    def lambda_endpoint(self, lambda_name: str) -> str:
+        return "https://" + self.api_lambda_domain(lambda_name)
+
+    def indexer_endpoint(self) -> str:
```
Comment (hannes-ucsc) [inline src/azul/__init__.py]: There are two places that would benefit from using this method:

https://github.com/DataBiosphere/azul/search?q=%22api_lambda_domain+indexer%22&unscoped_q=%22api_lambda_domain+indexer%22

File: src/azul/health.py [hunk 4]
```diff
@@ -1,40 +1,85 @@
...
+            except ClientError as ex:
+                response[queue] = {
+                    'up': False,
+                    'error': ex.response['Error']['Message']
+                }
+                response['up'] = False
+            else:
+                response[queue] = {
+                    'up': True,
+                    'messages':{
```
Comment (hannes-ucsc) [inline src/azul/health.py]:
```suggestion
                    'messages': {
```

File: test/health_check_test_case.py [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+
+    def _test(self, up: bool):
+        with ResponsesHelper() as helper:
+            helper.add_passthru(self.base_url)
+            for lambda_name in self._other_lambda_names():
+                helper.add(responses.Response(method='GET',
+                                              url=config.lambda_endpoint(lambda_name) + '/health/basic',
+                                              status=200 if up else 503,
+                                              json={'up': up}))
+            with mock.patch.dict(os.environ, AWS_DEFAULT_REGION="us-west-1"):
```
Comment (hannes-ucsc) [inline test/health_check_test_case.py]:
```suggestion
            with mock.patch.dict(os.environ, AWS_DEFAULT_REGION='us-east-1'):
```

File: test/indexer/test_health_check.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    @classmethod
+    def lambda_name(cls) -> str:
+        return 'indexer'
+
+
+# FIXME: This is inelegant: https://github.com/DataBiosphere/azul/issues/652
+del HealthCheckTestCase
+
+if __name__ == "__main__":
+    unittest.main()
```
Comment (hannes-ucsc) [inline test/indexer/test_health_check.py]: EOL missing at EOF

**Comment by hannes-ucsc** [comment]
Suggested commit title improvements:

> Move `memoized_property` to separate module

> Improve lambda endpoint configuration

> Improve health check endpoints (#624, #615, #548)

> Stop unittest from running abstract test cases (related to #652)

**Comment by hannes-ucsc** [comment]
A PR that resolves an issue doesn't block it.

**Comment by hannes-ucsc** [comment]
This is an externally visible change and the browser will have to integrate with it. This means we should also add an entry to our change log.

**Comment by achave11** [comment]
> A PR that resolves an issue doesn't block it isn't blocked by it.
> 
> [edited]

It was only a means of linking the PR with the issues. Since there is no direct PR addressing them.

**Comment by hannes-ucsc** [comment]
> It was only a means of linking the PR with the issues. Since there is no direct PR addressing them.

I understand the motivation but 1) the link direction should reflect reality and 2) I think this is confusing since we don't do this consistently.

File: src/azul/__init__.py [hunk 7]
```diff
@@ -290,6 +299,19 @@ def token_queue_name(self):
...
+    def fail_queue_name(self):
+        return config.qualified_resource_name('fail')
+
+    @property
+    def fail_fifo_queue_name(self):
+        return config.qualified_resource_name('fail', suffix='.fifo')
+
+    @property
+    def all_queue_names(self):
+        return (self.token_queue_name, self.document_queue_name, self.fail_queue_name,
```
Comment (hannes-ucsc) [inline src/azul/__init__.py]: Wrap all or none.

File: src/azul/health.py [hunk 8]
```diff
@@ -1,40 +1,85 @@
...
+        return {
+            'up': self.up,
+            'elastic_search': self.elastic_search,
+            'queues': self.queues,
+            **({
+                lambda_name: self._lambda(lambda_name)
+                for lambda_name in config.lambda_names()
+                if lambda_name != self.lambda_name
+            }),
+            'unindexed_bundles': sum(self.queues[config.notify_queue_name].get('messages',{}).values()),
```
Comment (hannes-ucsc) [inline src/azul/health.py]: This empty line should not be there.

**Pull request merged**